### ISSUE TITLE
preview fault with lightdm

### ISSUE THREAD

**Issue by sleeperzjf** [ranger/ranger#2023]
I'm using `archlinux` with `dwm`, and if I use `startx` to get into `dwm`, all my previews work all fine, but if I use `lightdm` to  get into `dwm`, the previews crashed.

**Comment by sleeperzjf** [comment]
when I use `lightdm`:
![image](https://user-images.githubusercontent.com/65698607/86109932-80dee080-baf7-11ea-9a72-4d16d811adb4.png)

**Comment by sleeperzjf** [comment]
When I use `startx`:
![image](https://user-images.githubusercontent.com/65698607/86110086-b388d900-baf7-11ea-8d7a-da990f245a6c.png)

**Comment by sleeperzjf** [comment]
I'm using `st` from `suckless` website. Has anyone encountered the same problem?

**Comment by toonn** [comment]
Does lightdm start a Wayland session for you? Ueberzug is X11-only.

**Comment by sleeperzjf** [comment]
No, I use newest X.

**Comment by toonn** [comment]
Would be great if you could test #2024. You should be able to clone the PR and then run `ranger.py` from the repo. It won't fix the problem yet but it should make the error reporting nicer.

**Comment by toonn** [comment]
You're not running ranger over ssh are you? I'm not sure how your DISPLAY would get messed up otherwise. [See here for the most common issue](https://unix.stackexchange.com/questions/199891/invalid-mit-magic-cookie-1-key-when-trying-to-run-program-remotely).

**Comment by toonn** [comment]
@seebye, it's proving really difficult to handle ueberzug's error reporting because it's asynchronous from ranger's POV. Is there a way to have ueberzug report on success? That way we could redirect stdout and stderr and just always do a blocking read. The only way we have right now of preventing ueberzug from messing up the UI is doing what amounts to redirecting stderr to /dev/null.

**Comment by seebye** [comment]
> it's proving really difficult to handle ueberzug's error reporting because it's asynchronous from ranger's POV. Is there a way to have ueberzug report on success?
> That way we could redirect stdout and stderr and just always do a blocking read.

Ueberzug works completely asynchonously and it's not really designed to give feedback.  
Also I spent quiet a time to make ueberzug fast which would be ruined by changing it's design to act synchonously.  
E.g. frame dropping wouldn't be possible  
Also this error is a crash not a reaction of executing a command.  
python-xlib fails to connect to X11.
If you want to log the output I think it's the best just to dump it to a file.

**Comment by kevinhwang91** [comment]
redirect stderr to PIPE. The user previews the second image, check the stderr first, if stderr exists, make the Ueberzug initialized failed.

**Comment by sleeperzjf** [comment]
No, I only installed `archlinux` on my laptop and use it for my daily
drive. As for the test, I don't get you, I pull the repo from `GitHub`
and run `ranger.py`, how do I make error reporting prettier?

**Comment by toonn** [comment]
@seebye, logging the output is only half the story. Currently we have no way to warn the user something's gone wrong. We can't wait for error output because it might never arrive, if we were assured there will always be output success or failure then we could wait.

@kevinhwang91, that's terrible UX. You're assuming people always preview multiple images one after the other. But the "next" image preview may never come or may be far removed in space or time. It may also come *before* any error output has been generated by ueberzug.

@sleeperzjf, nvm, the current PR isn't a good enough fix.

**Comment by kevinhwang91** [comment]
```diff
diff --git a/ranger/ext/img_display.py b/ranger/ext/img_display.py
index ffaa4c07..990fe607 100644
--- a/ranger/ext/img_display.py
+++ b/ranger/ext/img_display.py
@@ -735,15 +735,25 @@ class UeberzugImageDisplayer(ImageDisplayer):
             return
 
         self.process = Popen(['ueberzug', 'layer', '--silent'], cwd=self.working_dir,
-                             stdin=PIPE, universal_newlines=True)
+                             stdin=PIPE, stderr=PIPE, universal_newlines=True)
         self.is_initialized = True
+        timer_check_exit = threading.Timer(2, self._check_exit, [self.process])
+        timer_check_exit.start()
 
     def _execute(self, **kwargs):
         self.initialize()
         self.process.stdin.write(json.dumps(kwargs) + '\n')
         self.process.stdin.flush()
 
+    def _check_exit(self, process):
+        if process and  process.poll() == 1:
+            process.wait()
+            self.process = None
+
     def draw(self, path, start_x, start_y, width, height):
+        if self.is_initialized and self.process is None:
+            return
+
         self._execute(
             action='add',
             identifier=self.IMAGE_ID,
```

Create a timer delay 2s to check whether Ueberzug is exited. The result is what I expect. However, I only tested under python3.

**Comment by toonn** [comment]
@kevinhwang91, that's equally terrible UX. It'd freeze ranger's UI for 2 seconds after every image you move to *and* there's no guarantee 2 seconds is enough.

**Comment by kevinhwang91** [comment]
> @kevinhwang91, that's equally terrible UX. It'd freeze ranger's UI for 2 seconds after every image you move to _and_ there's no guarantee 2 seconds is enough.

Have you tried the above patch? Never block the main loop. The timer is asynchronous checking whether Ueberzug is exited. 

When the first Ueberzug process is checked to be exited, the draw method never initializes Ueberzug again.  If you think 2s delay time is not enough, you can set it to 10s or 1min, it doesn't matter. No matter the user preview how many images during the timer, the failed process will be closed finally.

**Comment by toonn** [comment]
@kevinhwang91, you're right I didn't look closely enough. What about errors that happen after initialization? And timers would still hold up previews for 2 seconds, which again might not be enough.

**Comment by kevinhwang91** [comment]
![rr](https://user-images.githubusercontent.com/17562139/86331111-35553f80-bc7b-11ea-8f7e-9f045603ca44.gif)
